File: launcher-tool/src/main/kotlin/LocalIO.kt
package dk.sdu.cloud

import java.io.File
import kotlin.system.exitProcess

class LocalFileFactory : FileFactory() {
    override fun create(path: String) = LocalFile(path)
}

class LocalFile(path: String) : LFile(path) {
    val jvmFile = File(path)

    override val absolutePath: String
        get() = jvmFile.absolutePath

    override fun exists(): Boolean = jvmFile.exists()

    override fun child(subpath: String): LFile {
        return LocalFile(File(jvmFile, subpath).absolutePath)
    }

    override fun writeBytes(bytes: ByteArray) {
        jvmFile.writeBytes(bytes)
    }

    override fun writeText(text: String) {
        jvmFile.writeText(text)
    }

    override fun appendText(text: String) {
        jvmFile.appendText(text)
    }

    override fun delete() {
        jvmFile.deleteRecursively()
    }

    override fun mkdirs() {
        jvmFile.mkdirs()
    }
}

val debugCommands = System.getenv("DEBUG_COMMANDS") != null

class LocalExecutableCommandFactory() : ExecutableCommandFactory() {
    override fun create(
        args: List<String>,
        workingDir: LFile?,
        postProcessor: (result: ProcessResultText) -> String,
        allowFailure: Boolean,
        deadlineInMillis: Long,
        streamOutput: Boolean,
    ) = LocalExecutableCommand(args, workingDir as LocalFile?, postProcessor, allowFailure, deadlineInMillis,
        streamOutput)
}

data class LocalExecutableCommand(
    override val args: List<String>,
    override val workingDir: LocalFile? = null,
    override val postProcessor: (result: ProcessResultText) -> String = { it.stdout },
    override var allowFailure: Boolean = false,
    override var deadlineInMillis: Long = 1000 * 60 * 5,
    override var streamOutput: Boolean = false,
) : ExecutableCommand {
    override fun toBashScript(): String {
        return buildString {
            if (workingDir != null) appendLine("cd '$workingDir'")
            appendLine(args.joinToString(" ") { "'${escapeBash(it)}'" })
        }
    }

    override fun executeToText(): Pair<String?, String> {
        if (debugCommands) println("Command: " + args.joinToString(" ") { "'$it'" })

        val deadline = System.currentTimeMillis() + deadlineInMillis

        val process = ProcessBuilder(args).apply {
            if (workingDir != null) {
                directory(workingDir.jvmFile)
            }
        }.start()

        process.outputStream.close()
        val output = process.inputStream.bufferedReader()
        val error = process.errorStream.bufferedReader()

        val outputBuilder = StringBuilder()
        val errBuilder = StringBuilder()
        val stdoutThread = Thread {
            while (System.currentTimeMillis() < deadline) {
                val line = output.readLine() ?: break
                if (streamOutput) printStatus(line)
                outputBuilder.appendLine(line)
            }
        }.also { it.start() }

        val stderrThread = Thread {
            while (System.currentTimeMillis() < deadline) {
                val line = error.readLine() ?: break
                if (streamOutput) printStatus(line)
                errBuilder.appendLine(line)
            }
        }.also { it.start() }

        stdoutThread.join()
        stderrThread.join()
        runCatching {
            output.close()
            error.close()

            if (process.isAlive) process.destroy()
        }

        val exitCode = process.waitFor()

        if (debugCommands) {
            println("  Exit code: ${exitCode}")
            println("  Stdout: ${outputBuilder}")
            println("  Stderr: ${errBuilder}")
        }

        if (exitCode != 0) {
            if (allowFailure) return Pair(null, outputBuilder.toString() + errBuilder.toString())

            println("Command failed!")
            println("Command: " + args.joinToString(" ") { "'$it'" })
            println("Directory: $workingDir")
            println("Exit code: ${exitCode}")
            println("Stdout: ${outputBuilder}")
            println("Stderr: ${errBuilder}")
            exitProcess(exitCode)
        }
        return Pair(postProcessor(ProcessResultText(exitCode, outputBuilder.toString(), errBuilder.toString())), "")
    }
}

File: launcher-tool-go/pkg/launcher/local_io.go
package launcher

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"ucloud.dk/launcher/pkg/termio"
)

type LocalFile struct {
	path string
}

func NewLocalFile(path string) LocalFile {
	return LocalFile{
		path: path,
	}
}

func (lf LocalFile) GetAbsolutePath() string {
	return lf.path
}

func (lf LocalFile) Exists() bool {
	_, err := os.Open(lf.path)
	if err != nil {
		return false
	} else {
		return true
	}
}

func (lf LocalFile) Child(subPath string, isDir bool) LFile {
	if isDir {
		err := os.MkdirAll(filepath.Join(lf.path, subPath), 0755)
		HardCheck(err)
		return NewLocalFile(filepath.Join(lf.path, subPath))
	} else {
		file, err := os.OpenFile(filepath.Join(lf.path, subPath), os.O_APPEND|os.O_CREATE|os.O_RDWR, 0666)
		HardCheck(err)
		return NewLocalFile(file.Name())
	}
}

func (lf LocalFile) WriteBytes(bytes []byte) {
	f, err := os.Create(lf.path)
	HardCheck(err)
	_, err = f.Write(bytes)
	SoftCheck(err)
}

func (lf LocalFile) WriteText(str string) {
	f, err := os.Create(lf.path)
	HardCheck(err)
	_, err = f.WriteString(str)
	SoftCheck(err)
}

func (lf LocalFile) AppendText(str string) {
	f, err := os.OpenFile(lf.path,
		os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	SoftCheck(err)
	defer f.Close()
	if _, err := f.WriteString(str); err != nil {
		log.Println(err)
	}
}

func (lf LocalFile) Delete() {
	err := os.RemoveAll(lf.path)
	HardCheck(err)
}

func (lf LocalFile) MkDirs() {
	err := os.MkdirAll(lf.path, 0755)
	HardCheck(err)
}

func (lf LocalFile) Name() string {
	return filepath.Base(lf.path)
}

type LocalExecutableCommand struct {
	args             []string
	workingDir       LFile
	fn               postProcessor
	allowFailure     bool
	deadlineInMillis int64
	streamOutput     bool
}

func NewLocalExecutableCommand(
	args []string,
	workingDir LFile,
	fn postProcessor,
) *LocalExecutableCommand {
	return &LocalExecutableCommand{
		args:             args,
		workingDir:       workingDir,
		fn:               fn,
		allowFailure:     false,
		deadlineInMillis: 1000 * 60 * 5,
		streamOutput:     false,
	}
}

func (l *LocalExecutableCommand) SetDeadline(deadlineInMillis int64) {
	l.deadlineInMillis = deadlineInMillis
}

func (l *LocalExecutableCommand) SetAllowFailure() {
	l.allowFailure = true
}

func (l *LocalExecutableCommand) SetStreamOutput() {
	l.streamOutput = true
}

func (l *LocalExecutableCommand) ToBashScript() string {
	sb := new(strings.Builder)
	if l.workingDir != nil {
		sb.WriteString("cd '" + l.workingDir.GetAbsolutePath() + "'\n")
	}
	for _, arg := range l.args {
		sb.WriteString(EscapeBash(arg) + " ")
	}
	sb.WriteString("\n")
	return sb.String()
}

func (l *LocalExecutableCommand) ExecuteToText() StringPair {
	if DebugCommandsGiven() {
		fmt.Println("Command: " + strings.Join(l.args, " "))
	}

	deadline := time.Now().UnixMilli() + l.deadlineInMillis

	dir := ""
	if l.workingDir != nil {
		dir = l.workingDir.GetAbsolutePath()
	}

	cmd := exec.Command(l.args[0], l.args[1:]...)
	cmd.Dir = dir

	stdout, err := cmd.StdoutPipe()
	SoftCheck(err)
	outputReader := bufio.NewReader(stdout)
	stderr, err := cmd.StderrPipe()
	SoftCheck(err)
	errorReader := bufio.NewReader(stderr)

	outputBuilder := &strings.Builder{}
	errBuilder := &strings.Builder{}

	err = cmd.Start()
	SoftCheck(err)

	wg := sync.WaitGroup{}

	wg.Add(1)
	go func() {
		defer wg.Done()
		for time.Now().UnixMilli() < deadline {
			str, err := outputReader.ReadString('\n')
			if len(str) == 0 && err != nil {
				break
			}
			if l.streamOutput {
				termio.Write("%s", str)
			}
			outputBuilder.Write([]byte(str))
		}
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()
		for time.Now().UnixMilli() < deadline {
			str, err := errorReader.ReadString('\n')
			if len(str) == 0 && err != nil {
				break
			}
			if l.streamOutput {
				termio.Write("%s", str)
			}
			errBuilder.Write([]byte(str))
		}
	}()

	err = cmd.Wait()
	SoftCheck(err)

	wg.Wait()

	exitCode := cmd.ProcessState.ExitCode()

	if DebugCommandsGiven() {
		fmt.Println("  Exit code ", strconv.Itoa(exitCode))
		fmt.Println("  Stdout ", outputBuilder.String())
		fmt.Println("  Stderr ", errBuilder.String())
	}

	if exitCode != 0 {
		if l.allowFailure {
			return StringPair{First: "", Second: outputBuilder.String() + errBuilder.String(), StatusCode: exitCode}
		}

		fmt.Println("Command failed!")
		fmt.Println("Command ", strings.Join(l.args, " "))
		fmt.Println("Directory: ", l.workingDir.GetAbsolutePath())
		fmt.Println("Exit code: ", strconv.Itoa(exitCode))
		fmt.Println("Stdout ", outputBuilder.String())
		fmt.Println("Stderr ", errBuilder.String())
		os.Exit(exitCode)
	}

	return StringPair{
		l.fn(
			ProcessResultText{
				statusCode: exitCode,
				stdout:     outputBuilder.String(),
				stderr:     errBuilder.String(),
			},
		),
		"",
		exitCode,
	}
}
